packages/cli/src/commands/dev/index.ts

```diff
@@ -48,6 +48,22 @@ const help = () => {
 };
 
 export default async function main(client: Client) {
+  if (process.env.__VERCEL_DEV_RUNNING) {
+    client.output.error(
+      `${cmd(
+        `${getPkgName()} dev`
+      )} must not recursively invoke itself. Check the Development Command in the Project Settings or the ${cmd(
+        'dev'
+      )} script in ${cmd('package.json')}`
+    );
+    client.output.error(
+      `Learn More: http://err.sh/vercel/recursive-invocation`
+    );
+    return 1;
+  } else {
+    process.env.__VERCEL_DEV_RUNNING = '1';
+  }
+
```

Comment:
This covers only the case when recursion happens in the dev command. The dev script recursion remains covered few lines below because to get rid of it we would need a change in the runtimes which is the only place which calls the dev script during `vercel dev`. Here https://github.com/vercel/runtimes/blob/a853c94419fb0ec934a8b0fb5dd84e0f016c281f/packages/static-build/src/index.ts#L520 we would need to exit the process when the child exits with an error. @AndyBitz mentioned that we are soon getting rid of that part of the builder so it's not necessary.